File: startup.pl
#!/usr/local/bin/perl

package nanobot;

# Look for threading support
sub detect_threads {
	$output->std("Checking for threading support ... ");
	eval {
		require threads;
		require threads::shared;
	};

	unless($@) {
		$ssl = 1;
		$output->good("[OK]\n");
	} else {
		$output->bad("[NO]\n");
		$output->bad("threads and threads::shared ara core requirements, please perl threads!\n");
		die;
	}
}

# Look for timer support
sub detect_timer {
	$output->std("Checking for timer support ....... ");
	eval {
		require Time::HiRes;
	};

	unless($@) {
		$ssl = 1;
		$output->good("[OK]\n");
	} else {
		$output->bad("[NO]\n");
		$output->bad("Time::HiRes is a core requirement, please install it!\n");
		die;
	}
}

# Look for module support
sub detect_module_load {
	$output->std("Checking for module support ...... ");
	eval {
		require Module::Load;
	};

	unless($@) {
		$ssl = 1;
		$output->good("[OK]\n");
	} else {
		$output->bad("[NO]\n");
		$output->bad("Module::Load is a core requirement, please install it!\n");
		die;
	}
}

# Look for IPv6 support
sub detect_ipv6 {
	$output->std("Checking for IPv6 support ........ ");
	$ipv6 = 0;
	eval {
		require IO::Socket::INET6;
	};

	unless($@) {
		$ipv6 = 1;
		$output->good("[OK]\n");
		require "connect6.pl"
	} else {
		$output->bad("[NO]\n");
		require "connect.pl"
	}
}

# Look for SSL support
sub detect_ssl {
	$output->std("Checking for SSL support ......... ");
	$ssl = 0;
	eval {
		require IO::Socket::SSL;
	};

	unless($@) {
		$ssl = 1;
		$output->good("[OK]\n");
	} else {
		$output->bad("[NO]\n");
	}
}
1;

File: startup.rb
#!/usr/bin/ruby

# Class to do inital checks and load available libraries
class Startup
	def initialize( status, config, output )
		@status	= status
		@config	= config
		@output	= output
	end

	def checksocket
		@output.std( "Checking for socket libs ......... " )
		begin
			require 'socket'
			@output.good( "[OK]\n" )
		rescue LoadError
			@output.bad( "[NO]\n" )
			@output.info( "Ruby socket library needs to be installed.\n" )
			Process.exit
		end
	end

	def checkopenssl
		@output.std( "Checking for openssl support ..... " )
		begin
			require 'openssl'
			@output.good( "[OK]\n" )
			@status.ssl( 1 )
		rescue LoadError
			@output.bad( "[NO]\n" )
			@status.ssl( 0 )
		end
	end

	def checkthreads
		@output.std( "Checking for threading support ... " )
		begin
			require 'thread'
			@status.threads( 1 )
			@output.good( "[OK]\n" )
		rescue LoadError
			@status.threads( 0 )
			@output.bad( "[NO]\n" )

			if( !@config.threadingfallback )
				@output.info( "No threading support, and fallback is disabled.\n" )
				Process.exit
			end
		end
	end
end
